### PR TITLE
2020 12 01 secretbox

### PR DESCRIPTION
**Pull request by thedavidmeister**
implements xsalsa20poly1305 a.k.a. 'secretbox' from libsodium as a host fn

things for the future:

- better handling of sensitive memory in wasm
- using actual libsodium instead of a rust crate
- 'box' for keypair based encryption
- wasi crypto api or similar

### CODE REVIEW

File: crates/zome_types/src/lib.rs [hunk 1]
```diff
@@ -73,3 +75,121 @@ pub trait CallbackResult {
...
+///  - The host/rust generally doesn't guarantee to immediately wipe/zero out freed memory, either
+///    when a zome call is running or after a wasm instance is thrown away.
+///
+/// Most of the time we should just try to minimise the interaction between wasm and secret data.
+///
+/// For example, lair keeps all our private keys internal and we can only send it signing requests
+/// associated with public keys.
+///
+/// In other contexts it is more difficult, such as when generating secrets from raw cryptographic
+/// random bytes and sending them to peers directly.
```
Comment (neonphog) [inline crates/zome_types/src/lib.rs:131]: From an infosec perspective, you should just never do this. This is the point of asynchronous crypto - individuals generate keypairs and keep their private keys private, only publishing public keys.

I worry about publishing an API that encourages bad practices.

File: crates/zome_types/src/lib.rs [hunk 2]
```diff
@@ -73,3 +75,121 @@ pub trait CallbackResult {
...
+/// e.g. We don't want other machines to simply `remote_call` a successful timing attack.
+///
+/// MITM attacks are mitigated by the networking implementation itself.
+///
+/// @todo given how impossible it is for wasm to protect its memory from the host, it would make
+/// more sense to:
+///
+///  - use key exchange protocols like libsodium kx https://libsodium.gitbook.io/doc/key_exchange.
+///  - keep secrets inside lair with all algorithms behind an API, wasm only has access to opaque
+///    references to the secret data.
```
Comment (neonphog) [inline crates/zome_types/src/lib.rs:143]: Either / both of these are totally valid strategies to base our API on - we don't even need lair to be implemented yet, we can keep the private keys unprotected in holochain for the moment, but write the WASM api properly to interact with handles or key exchange mechanisms so app devs don't go down an insecure path with their code, right?

**Review by neonphog — CHANGES_REQUESTED** [review]
From some of the comments here, I'm getting the impression you are trying to effect a membership-based encryption where a curated group is given access keys to decrypt publicly available messages?

Would what you are trying to accomplish be better supported by a classic key ring implementation? Like, the symmetric key is asymmetrically encrypted separately for each member of the group / set of messages, and membership is the act of an existing member re-encrypting that key for that agent's public key?

I'm not meaning to shut you down, totally just trying to have a conversation before we bake a behavior into our API.

**Comment by thedavidmeister** [comment]
> From some of the comments here, I'm getting the impression you are trying to effect a membership-based encryption where a curated group is given access keys to decrypt publicly available messages?
> 
> Would what you are trying to accomplish be better supported by a classic key ring implementation? Like, the symmetric key is asymmetrically encrypted separately for each member of the group / set of messages, and membership is the act of an existing member re-encrypting that key for that agent's public key?
> 
> I'm not meaning to shut you down, totally just trying to have a conversation before we bake a behavior into our API.

@neonphog yes, that's why i put "'box' for keypair based encryption" as the next item on the list, so you can box up a secretbox key, e.g. how saltpack does it

we need to _have_ a symmetric key before we can encrypt it though :sweat_smile: 

> Either / both of these are totally valid strategies to base our API on - we don't even need lair to be implemented yet, we can keep the private keys unprotected in holochain for the moment, but write the WASM api properly to interact with handles or key exchange mechanisms so app devs don't go down an insecure path with their code, right?

@neonphog both are useful in different contexts, so it would be good to have them both

i think when i wrote this i was thinking that it would be too limiting to _force_ it to work this way but now that i've slept on it and have a fresh coffee, i'd like to design something that works better here :coffee:

**Review by neonphog — APPROVED** [review]
Nice!

**Comment by thedavidmeister** [comment]
over here for the combined PR :) - https://github.com/holochain/holochain/pull/559

**Pull request closed without merging**